src/openms/source/FILTERING/DATAREDUCTION/Deisotoper.cpp

```diff
@@ -38,9 +38,285 @@
+  std::vector<Size> iso_peak_count(old_spectrum.size(), 1);
+  int feature_number = 0;
+
+  std::vector<size_t> extensions;
+  std::vector< std::vector<size_t> > clusters;
+  std::vector<int> charges_of_extensions;
+  const float averagine_check_threshold[5] = { 0.05f,0.1f,0.2f,0.4f,0.6f };
+
+  bool has_precursor_data(false);
+  double precursor_mass(0);
+  if (old_spectrum.getPrecursors().size() == 1)
+  {
+    has_precursor_data = true;
+    int precursor_charge = old_spectrum.getPrecursors()[0].getCharge();
+    precursor_mass = (old_spectrum.getPrecursors()[0].getMZ() * precursor_charge) - (Constants::PROTON_MASS * precursor_charge);
+  }
+
+  for (size_t current_peak = 0; current_peak != old_spectrum.size(); ++current_peak)
+  {
+    // only process peaks which are not already in a cluster. Would form clusters identical to the cluster they are assigned to,
+    // excluding its first peak(s), since peaks with lower mz are not considered for cluster-formation.
+    if (features[current_peak] != -1) continue;
+
+    // Monoisotopic peaks with intensity 0.0 interfere with averagine check when normalizing the spectrum peaks height
+    if (use_averagine_model && !rem_low_intensity && old_spectrum[current_peak].getIntensity() == 0.0) continue;
+
+    const double current_mz = old_spectrum[current_peak].getMZ();
+    if (add_up_intensity)
+    {
+      mono_iso_peak_intensity[current_peak] = old_spectrum[current_peak].getIntensity();
+    }
+    clusters.clear();
+    charges_of_extensions.clear();
+    for (int q = max_charge; q >= min_charge; --q) // important: test charge hypothesis from high to low
+    {      
+      // try to extend isotopes from mono-isotopic peak
+      // if appropriate extension larger than min_isopeaks possible:
+      //   - save charge q in mono_isotopic_peak[]
+      //   - annotate_charge all isotopic peaks with feature number
+
+      bool has_min_isopeaks = true;
+      const double tolerance_dalton = fragment_unit_ppm ? Math::ppmToMass(fragment_tolerance, current_mz) : fragment_tolerance;
+
+      // do not bother testing charges q (and masses m) with: m/q > precursor_mass/q (or m > precursor_mass)
+      if (has_precursor_data)
+      {
+        double current_theo_mass = (current_mz * q) - (Constants::PROTON_MASS * q);
+        if (current_theo_mass > (precursor_mass + tolerance_dalton))
+        {
+          continue;
+        }
+      }
+
+      extensions.clear();
+      extensions.push_back(current_peak);
+
+      float total_weight = old_spectrum[current_peak].getIntensity() * (old_spectrum[current_peak].getMZ() * q - (q - 1) * Constants::PROTON_MASS_U);
+      float total_intensity = old_spectrum[current_peak].getIntensity();
+      for (unsigned int i = 1; i < max_isopeaks; ++i)
+      {
+        const double expected_mz = current_mz + static_cast<double>(i) * Constants::C13C12_MASSDIFF_U / static_cast<double>(q);
+        const int p = old_spectrum.findNearest(expected_mz, tolerance_dalton);
```

Comment:
does the original publication only look for the nearest peak or consider all peaks in the tolerance window?

Reply:
"A set of possible isotopic clusters is then obtained by checking the m/z distance between peaks, considering fragment charge states [...]

Does not really say. We could try to pick the peak in the tolerance window that achieves the best KL. Will give it a look as well.